lsp-completion.el

```diff
@@ -0,0 +1,542 @@
+
+        (setq-local lsp-inhibit-lsp-hooks nil)
+
+        (when (lsp-completion--looking-back-trigger-characterp trigger-chars)
+          (setq this-command 'self-insert-command)))
+    (lsp-completion--capf-clear-cache)))
+
+;;;###autoload
+(defun lsp-completion--enable ()
+  "Enable LSP completion support."
+  (when (and lsp-completion-enable
+             (lsp-feature? "textDocument/completion"))
+    (lsp-completion-mode 1)))
+
+(defun lsp-completion--disable ()
+  "Disable LSP completion support."
+  (lsp-completion-mode -1))
+
+;;;###autoload
+(define-minor-mode lsp-completion-mode
+  "Toggle LSP completion support."
+  :group 'lsp-mode
+  :global nil
+  :lighter ""
+  (cond
+   (lsp-completion-mode
+    (setq-local completion-at-point-functions nil)
+    (add-hook 'completion-at-point-functions #'lsp-completion-at-point nil t)
+    (setq-local completion-category-defaults
+                (add-to-list 'completion-category-defaults '(lsp-capf (styles basic))))
+
+    (when (lsp--capability :completionProvider)
+      (cond
+       ((and (functionp 'company-lsp)
+             (not lsp-completion-prefer-capf))
+        (progn
+          (company-mode 1)
+          (add-to-list 'company-backends 'company-lsp)
+          (setq-local company-backends (remove 'company-capf company-backends))))
+
+       ((and (fboundp 'company-mode) lsp-completion-enable)
+        (company-mode 1)
+        (add-to-list 'company-backends 'company-capf))))
+
+    (when (bound-and-true-p company-mode)
+      (add-hook 'company-completion-started-hook
+                (lambda (&rest _)
+                  (setq-local lsp-inhibit-lsp-hooks t))
+                nil
+                t)
+      (add-hook 'company-after-completion-hook
+                (lambda (&rest _)
+                  (lsp-completion--capf-clear-cache)
+                  (setq-local lsp-inhibit-lsp-hooks nil))
+                nil
+                t))
+    (add-hook 'lsp-configure-hook #'lsp-completion--enable nil t)
+    (add-hook 'lsp-unconfigure-hook #'lsp-completion--disable nil t))
+   (t
+    (remove-hook 'completion-at-point-functions #'lsp-completion-at-point t)
+    (setq-local completion-category-defaults
+                (cl-remove 'lsp-capf completion-category-defaults :key #'car))
+    (remove-hook 'lsp-configure-hook #'lsp-completion--enable t)
+    (remove-hook 'lsp-unconfigure-hook #'lsp-completion--disable t)
+    (when (featurep 'company)
+      (remove-hook 'company-completion-started-hook
+                   (lambda (&rest _)
+                     (setq-local lsp-inhibit-lsp-hooks t))
+                   t)
+      (remove-hook 'company-after-completion-hook
+                   (lambda (&rest _)
+                     (lsp-completion--capf-clear-cache)
+                     (setq-local lsp-inhibit-lsp-hooks nil))
+                   t)))))
+
+;;;###autoload
+(add-hook 'lsp-configure-hook (lambda ()
+                                (when lsp-completion-enable
```

Comment:
seems like it should be `(when (and lsp-auto-configure lsp-completion-enable) ...` if we want to have one place to turn everything off.

Reply:
Currently `lsp-configure-hook` is only run when `lsp-auto-configure` so it's not needed now. I think we can defer this until we fix how `lsp-auto-configure` affect `lsp-configure-hook`, that will be one swoop change